src/change_res.cpp

```diff
@@ -68,16 +74,176 @@ void ChangeRes(int, int, int, int)
 
 void UpdateInternalRes()
 {
+    int req_w = g_config.InternalW;
+    int req_h = g_config.InternalH;
+
+    if(!g_compatibility.allow_multires)
+    {
+        // TODO: use the correct canonical screen's resolution here
+        if((req_w != 0 && req_w < 800) || (req_h != 0 && req_h < 600))
+        {
+            req_w = 800;
+            req_h = 600;
+        }
+    }
+
+    if(req_w == 0 || req_h == 0)
+    {
+        int int_w, int_h, orig_int_h;
+
+        XRender::getRenderSize(&int_w, &int_h);
+        orig_int_h = int_h;
+
+        // set internal height first
+        if(req_h != 0)
+        {
+            int_h = req_h;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_FIXED_05X)
+        {
+            int_h *= 2;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_FIXED_2X)
+        {
+            int_h /= 2;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_DYNAMIC_INTEGER)
+        {
+            if(int_h >= 600)
+            {
+                // constrains height to be in the 600-720p range
+                int scale_factor = int_h / 600;
+                int_h /= scale_factor;
+            }
+        }
+
+        // minimum height constraint
+        if(int_h < 320)
+            int_h = 320;
+
+        // maximum height constraint
+        if(int_h > 720 && req_h <= 720)
+            int_h = 720;
+
+        // now, set width based on height and scaling mode
+        if(g_videoSettings.scaleMode == SCALE_FIXED_05X)
+        {
+            int_w = int_w * 2;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_FIXED_1X)
+        {
+            int_w = int_w;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_FIXED_2X)
+        {
+            int_w = int_w / 2;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_DYNAMIC_INTEGER)
+        {
+            int scale_factor = orig_int_h / int_h;
+            if(scale_factor == 0)
+            {
+                int_w = int_w;
```

Comment:
Why there is an assignment to self? You may want just put a comment here, otherwise compilers may drop a warning on this.